### DISCUSSION TITLE
tutorials/inputs/input_examples

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#287] [User-contributed notes]
# tutorials/inputs/input_examples

Introduction: In this tutorial, you'll learn how to use Godot's InputEvent system to capture player input. There are many different types of input your game may use - keyboard, gamepad, mouse, etc....

https://docs.godotengine.org/en/stable/tutorials/inputs/input_examples.html

**Comment by bretgourdie** [comment]
I spent quite a bit of time figuring out that I needed the `allowEcho` argument so my `InputEvent.IsActionPressed(string actionName)` call wouldn't report that I am no longer holding a keyboard key after about one second. It's not mentioned on this page at all, and it is barely discussed in `InputEvent` (https://docs.godotengine.org/en/stable/classes/class_inputevent.html#class-inputevent-method-is-action-pressed). I'm surprised it doesn't trip more people up.

**Reply by Calinou** [reply]
Can you show the full code sample where you're using `IsActionPressed`? You shouldn't need to change this argument from its default in most situations. Usually, only non-game applications need to care about it.

**Reply by gkushalg01** [reply]
while I agree that it should have been discussed, I also took some time in understanding. btw did you try is_action_released for your purpose?

**Comment by Jeremi360** [comment]
From me this doesn't work: `if event.is_action_pressed(&"select):`
But this works:
```
if event.is_action(&"select"):
			if event.is_pressed():
```

**Reply by Calinou** [reply]
Which input are you testing with? `is_action_pressed()` doesn't work with mouse wheel inputs, as they are both pressed and released on the same frame when they occur. (This is the only input that works like this.)

**Reply by Jeremi360** [reply]
it didn't work for "mouse left button", "ctrl" and "shift".

**Reply by matthew-dean** [reply]
I've also been struggling to get mouse clicks working correctly. It doesn't trigger the action either the same way as the keyboard would or it does so only after both a press and release, so it is perceivably delayed.

**Comment by BadlyRaisin** [comment]
I ran into unexpected behavior with InputActions and thought I would share in case anyone else has a similar issue. By default, ui_undo is mapped to ctrl+z while ui_redo is mapped to ctrl+shift+z and ctrl+y. I couldn't get ui_redo to trigger with ctrl+shift+z, but it worked fine with ctrl+y. Eventually I realized it was because ui_undo was triggering first since I was hitting ctrl+z (and shift). I solved the immediate problem by checking for ui_redo first and not checking for undo if redo was found. This made me realize that I will need some additional checks if I allow the player to rebind controls since they might create a similar overlap even without having two actions be mapped to explicitly identical bindings.

Problem code (since ctrl+shift+z includes ctrl+z):
```
...
elif event.is_action_pressed("ui_undo"):
	undo_redo.undo()
elif event.is_action_pressed("ui_redo"):
	undo_redo.redo()
```
...
Fixed (check for more specific combination first):
```
elif event.is_action_pressed("ui_redo"):
	undo_redo.redo()
elif event.is_action_pressed("ui_undo"):
	undo_redo.undo()
```

**Reply by Calinou** [reply]
`InputEvent.is_action_pressed()` has an optional `exact_match`, which you should set to `true` for situations like these. It ensures that shortcuts that use *more* modifiers than needed don't match. In this case, pressing <kbd>Ctrl + Shift + Z</kbd> won't match `ui_undo` if `exact_match` is `true`.

```gdscript
# "exact_match" is the last parameter.
elif event.is_action_pressed("ui_undo", false, true):
	undo_redo.undo()
elif event.is_action_pressed("ui_redo", false, true):
	undo_redo.redo()
```
